core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -857,9 +862,11 @@ private[spark] class BlockManager(
 
     val startTimeMs = System.currentTimeMillis
     var blockWasSuccessfullyStored: Boolean = false
+    var exceptionWasThrown: Boolean = true
     val result: Option[T] = try {
       val res = putBody(putBlockInfo)
       blockWasSuccessfullyStored = res.isEmpty
+      exceptionWasThrown = false
       res
     } finally {
       if (blockWasSuccessfullyStored) {
```

Comment:
One concern with using a `catch` here is handling of `InterruptedException`: if we use `case NonFatal(e)` that won't match `InterruptedException` and we'll miss out on cleanup following that. If we catch `Throwable`, on the other hand, then I think that we'll end up clearing the `isInterrupted` bit for `InterruptedException`s and it'll be awkward to match and re-set it when rethrowing. Therefore I'd like to keep the exception-handling case in the `finally` block with a simple check to see if we entered that block via an error case.

Note that I've seen this same exception-handling idiom used in Java code, where code that catches and re-throws `Throwable` won't compile in older Java versions because of the checked exception-handling (I think that newer versions are a bit more permissive about throwing exceptions from a `catch` block).